openupgradelib/openupgrade_90.py

```diff
@@ -66,3 +68,29 @@ def convert_binary_field_to_attachment(env, field_spec):
             env.cr.execute("ALTER TABLE {} DROP COLUMN {}".format(
                 model._table, column,
             ))
+
+
+def replace_account_types(env, type_spec, unlink=True):
+    """ Replace old account types with their replacements.
+    :param type_spec: list of tuples (xmlid of old account.account.type, \
+xmlid of new account.account.type)
+    :param unlink: attempt to unlink the old account type
+    """
+    logger = logging.getLogger('OpenUpgrade')
+    for old_type, new_type in type_spec:
+        type8 = env.ref(old_type)
```

Comment:
Maybe one of these types has been deleted on the target DB, so you should make:
```
type8 = env.ref(old_type, False)
if not type8:
    continue
```

Reply:
I prefer not to allow for this. The account types should exist because they are not defined with FORCECREATE=0. If they don't exist the user can upgrade the modules on the source version of the migration first. If we don't allow for non-existing types there is an implicit check on the correct spelling of the xmlids.

Reply:
But this is a general utility. You can use this method for custom or other uses than the one you have thought, so it's better to not block migrations by this detail. Put a warning if you want to be verbose, but don't make this to fail.

Reply:
No, I want it to fail. Leave it up to the caller to only pass valid account types. For the record, are there even any account types in all of the Odoo localization modules that are defined with forcecreate=false?

Reply:
Answering my own question: no there are no account types defined with forcecreate=false in the Odoo localization modules.

Reply:
I'm thinking in the case you haven't updated previous version before starting migration (that's not a requirement till now), and you removed one of the types.

Reply:
You are right that it is not a formal requirement at the moment. Do you think it is reasonable to add it as a requirement? It is not just for this particular case. There may be other cases which would make it easier for the developers if module data can be considered consistent.

Reply:
I would prefer to avoid it for this special case, as the check code is very simple, and till now I haven't found any other thing that requires to be fully updated, but we can include at least a reference on documentation.

Reply:
Then I'll modify the code here and only raise for missing records if the testing flag is present on the thread.